lnrpc/routerrpc/router.proto

```diff
@@ -490,6 +490,136 @@ message BuildRouteResponse {
     lnrpc.Route route = 1;
 }
 
+
+message SubscribeHtlcEventsRequest{}
+
+message HtlcEvent{
+    /**
+    HtlcKey identifies htlcs and can be used to match forward events to their
+    corresponding settle or fail event.
+    */
+    HtlcKey htlc_key = 1 [json_name="htlc_key"];
+
+    /**
+    Timestamp is the time (unix epoch offset in seconds) that the event
+    occurred.
+    */
+    uint64 timestamp = 2 [json_name="timestamp"];
+
+    enum EventType{
+        UNKNOWN = 0;
+        SEND = 1;
+        RECEIVE = 2;
+        FORWARD = 3;
+    }
+
+    /**
+    The event type indicates whether the htlc was part of a send, receive or
+    forward.
+    */
+    EventType event_type = 3 [json_name ="event_type"];
+
+    oneof event {
+        ForwardEvent forward_event = 4 [json_name = "forward_event"];
+        ForwardFailEvent forward_fail_event = 5 [json_name = "forward_fail_event"];
+        SettleEvent settle_event = 6 [json_name = "settle_event"];
+        LinkFailEvent link_fail_event = 7 [json_name= "link_fail_event"];
+    }
+}
+
+message HtlcKey{
+    /**
+    IncomingChannel is the channel that the incoming htlc arrived at our node
+    on. This value is zero for sends.
+    */
+    uint64 incoming_channel = 1 [json_name="incoming_channel"];
+
+    /**
+    Outgoing channel is the channel that the outgoing htlc left our node on.
+    This value is zero for receives.
+    */
+    uint64 outgoing_channel = 2 [json_name="outgoing_channel"];
+
+    /**
+    Incoming id is the index of the incoming htlc in the incoming channel.
+    This value is zero for sends.
+    */
+    uint64 incoming_htlc_id = 3 [json_name="incoming_htlc_id"];
+
+    /**
+    Outgoing id is the index of the outgoing htlc in the outgoing channel.
+    This value is zero for receives.
+    */
+    uint64 outgoing_htlc_id = 4 [json_name="outgoing_htlc_id"];
+}
+
+message HtlcInfo{
```

Comment:
I know we have the split between `HtlcInfo` and `HtlcKey` because some available isn't accessible when the event is generated. I think it is ok internally, but do we also want to do that on the external rpc interface? It would promote something that really is 'not yet implemented' to a factor that shapes the response structure in a way that isn't easily undone.

Reply:
Decided to keep this separate from for `LinkFail`/`Forwarding` events, because I'm not sure it's worth a change to the switch just to add these values for rpc purposes. If we decide to add them, htlc info fields can be deprecated and moved into the main `HtlcEvent` struct.

Reply:
Ok. The message is marked experimental, so we have the opportunity to change it later.